File: app/src/main/java/com/channa/mobiledatausageapp/di/module/DataModule.java
package com.channa.mobiledatausageapp.di.module;

import com.channa.mobiledatausageapp.network.APIClient;
import com.channa.mobiledatausageapp.repository.MobileDataRepository;
import com.channa.mobiledatausageapp.viewmodel.CustomViewModelFactory;

import javax.inject.Singleton;

import androidx.lifecycle.ViewModelProvider;
import dagger.Module;
import dagger.Provides;

@Module
public class DataModule {

    @Provides
    @Singleton
    MobileDataRepository provideMobileDataRepository(APIClient apiClient) {
        return new MobileDataRepository(apiClient);
    }

    @Provides
    @Singleton
    ViewModelProvider.Factory provideViewModelFactory(MobileDataRepository mobileDataRepository) {
        return new CustomViewModelFactory(mobileDataRepository);
    }

}

File: app/src/main/java/com/zak/datauses/di/module/DataModule.kt
package com.zak.datauses.di.module

import androidx.lifecycle.ViewModelProvider
import com.zak.datauses.network.APIClient
import com.zak.datauses.repository.MobileDataRepository
import com.zak.datauses.viewmodel.ViewModelFactoryMobile
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class DataModule {
    @Provides
    @Singleton
    fun provideMobileDataRepository(apiClient: APIClient?): MobileDataRepository {
        return MobileDataRepository(apiClient!!)
    }

    @Provides
    @Singleton
    fun provideViewModelFactory(mobileDataRepository: MobileDataRepository?): ViewModelProvider.Factory {
        return ViewModelFactoryMobile(mobileDataRepository!!)
    }
}